### DISCUSSION TITLE
udp保持连接上有可怕的bug

### DISCUSSION THREAD

**Discussion by harvey-git** [opengnb/opengnb#27] [General]
开启服务后，udp连接无限增加，我家的路由器器默认设置是net.nf_conntrack_max = 40960 ，大约十分钟就会满载，无论如何减少timout相关参数都没用，最初没有发现是这个导致的，于是增加 net.nf_conntrack_max 的值，然而无论增加多少，都会迅速满载，我甚至试过增加到1048576，仍旧能够在一小会儿后满载，这么大的连接数量瞬间干掉了我tproxy的xray，更是惊动办公室的网管拿着公司防火墙日志来找我询问我的办公室电脑在干嘛……我折腾了xray数天后，才开始考虑这个异常数量的udp来源，在分别停掉tailscale，zerotier以及n2n后，都没有变化，最后才忍痛停掉opengnb后，udp的conntrack数量开始下降，20分钟后恢复正常水平(我的timeout最长就是20分钟)我才确定是这玩意儿有贪婪属性，注意，我后来反复测试过，这东西只会在两个机器之间通信过之后才会疯狂建立udp连接数量，然后如果这机器同时连接的另一端一多(我最多同时部属在6台机器上)，只要分别ssh一下其他机器，即使再断开所有ssh进程，跟这些机器之间的udp连接数量就会疯狂增长，而几乎不会有减少。这是一个非常强大的穿透工具，我个人测试穿透能力远远强于上面提到的另外三个同类软件，希望能修复这个bug。

**Comment by gnbdev** [comment]
你反映的这个情况在去年的时候已经在GNB技术支持的微信群里已经有结论了，我的结论不认为这是一个bug，但是确实需要对这个问题作改善。不确定你是否在微信群，可以加我微信拉你进群，QQ微信同号26253。
NAT穿透需要两端的节点进行地址和端口的碰撞，依据经验碰撞的面大且快成功率就高，但是这样对节点所在的路由防火墙确实造成一些不必要的负载，尤其是硬件性能较低的网关。目前有用户自行修改代码去改善，今后的版本会提供更多的启动选项让用户去调整节点探测的频率，我现在在忙别的项目，如果你可以为gnb写一些入门的教程或一些视频发布在自媒体平台，这会让我尽快回到gnb这个项目来。

UDP没有“连接”的概念，这段描述不符合常理，“只要分别ssh一下其他机器，即使再断开所有ssh进程，跟这些机器之间的udp连接数量就会疯狂增长，而几乎不会有减少”，没有更多信息情况下只能认为这与 GNB 无关，要查证非常简单，tcpdump，gnb 数据包很精且加密，但首部有明显的特征且是包含一些辨识度很高的文本信息方便用户判断这些数据包的类型及GNB正在做什么。
